recidiviz/persistence/persistence.py

```
@@ -16,27 +16,41 @@
 # =============================================================================
 """Contains logic for communicating with the persistence layer."""
 
+from recidiviz import Session
+from recidiviz.persistence import entity_matching
+from recidiviz.persistence.database.schema import Person
 
-class Persistence(object):
-    def __init__(self, database):
-        """
-        Setup persistence communication using the provided Database.
-
-        :param database: The database to persist into.
-        """
-        self._database = database
-
-    def write(self, ingest_info):
-        """
-        Persist each person in the ingest_info. If a person with the given
-        surname/birthday already exists, then skip that person.
-
-        :param ingest_info: The IngestInfo containing the surname.
-        """
-        for person in ingest_info.person:
-            person_already_exists = self._database.read_people(person.surname,
-                                                               person.birthdate)
-            if person_already_exists:
-                continue
-
-            self._database.write_person(person.surname, person.birthdate)
+
+def write(ingest_info):
+    """
+    Persist each person in the ingest_info. If a person with the given
+    surname/birthday already exists, then update that person.
+
+    :param ingest_info: The IngestInfo containing each person.
+    """
+    for ingest_info_person in ingest_info.person:
+        person = _convert_person(ingest_info_person)
+        session = Session()
+        try:
+            existing_person = entity_matching.get_entity_match(person, session)
+
+            if existing_person is None:
+                session.add(person)
+            else:
+                person.person_id = existing_person.person_id
+                session.add(session.merge(person))
+
+            session.commit()
+        except Exception as e:
```

:+1: I like this idea of letting exceptions bubble up from the database layer into the persistence layer to be handled here, making sure not to let them bubble further into the app layer.

---

As written, this should rethrow/reraise the exception. Do you mean it would make more sense to:
1. Catch the exception here 
2. Maybe log it?
3. Rollback the transaction
4. Just return like normal? Or maybe return a boolean for if the write was successful?

I guess this question really boils down to: How do you imagine the persistence layer being used and if that code has different logic if a ingest_info write fails?

---

No, I like it as-is here. This was meant only as a compliment for your current approach.  I think the persistence layer is the layer that our ingest app will speak to to persist entities and retrieve persisted entities. I think the database layer should generally not catch its own exceptions but let them bubble up to the persistence layer to be handled here.

I think it's then debatable as to whether the persistence layer should catch exceptions and propagate them (as you have here), or catch exceptions and then return certain error signals. Given that exceptions while communicating with the database are true errors and should be expected to be transient and should be retried, I think propagating the exception is the right move, probably.

An issue you'd likely be interested in is #114: should we continue this return -1 up the stack to the handler where it's converted into an error response to provoke retry? or should we pursue a different approach?